engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/IOMappingHelper.java

```diff
@@ -45,34 +48,56 @@
       variablesState.removeTemporaryVariables(elementInstanceKey);
     }
 
-    if (hasOutputMappings) {
-      mergeTool.reset();
-
-      final DirectBuffer variables =
-          determineVariables(variablesState, elementInstanceKey, outputMappings);
-
-      mergeTool.mergeDocumentStrictly(variables, outputMappings);
-      final DirectBuffer mergedVariables = mergeTool.writeResultToBuffer();
+    if (!outputMappings.isEmpty()) {
+      final DirectBuffer mergedVariables =
+          mergeOutputVariables(variablesState, elementInstanceKey, outputMappings);
 
       variablesState.setVariablesFromDocument(
           getVariableScopeKey(context), workflowKey, mergedVariables);
     }
   }
 
+  private DirectBuffer mergeOutputVariables(
+      final VariablesState variablesState,
+      final long elementInstanceKey,
+      final Mappings outputMappings) {
+
+    mergeTool.reset();
+
+    // fetch the variables from the target mappings to allow merging with existing variables
+    final DirectBuffer targetVariables =
+        variablesState.getVariablesAsDocument(
+            elementInstanceKey, outputMappings.getTargetVariableNames());
+
+    // copy variables to not override them with next variable state access
+    buffer.putBytes(0, targetVariables, 0, targetVariables.capacity());
+    bufferView.wrap(buffer, 0, targetVariables.capacity());
+
+    mergeTool.mergeDocument(bufferView);
+
+    final DirectBuffer sourceVariables =
+        variablesState.getVariablesAsDocument(
+            elementInstanceKey, outputMappings.getSourceVariableNames());
+
+    mergeTool.mergeDocumentStrictly(sourceVariables, outputMappings.get());
```

Comment:
Do you remember why this is called strictly ? :sweat_smile:

Reply:
I checked it. `Strictly` means here that an exception is thrown when a mapping cannot be applied (i.e. no variable or nested property with the given name). The other method doesn't fail in this case. But it seems that we don't use the non-strict version of this anywhere :thinking:

Reply:
Ah ok thanks. Maybe we can delete it in a separate issue?

Reply:
Sure. I'll try to remove it :wastebasket: